File: 2023/kentremi/src/days/day1.ts
import { FileHandle } from 'fs/promises';
import { wordToNumber } from '../utils';

async function day1(task: number, dayFileHandle: FileHandle) {
  let result = 0;
  for await (const line of dayFileHandle.readLines()) {
    const matchEx =
      task === 1
        ? /\d/g
        : /(?=(one|two|three|four|five|six|seven|eight|nine|\d))/g;

    const numbersInString = [];

    let match;
    while ((match = matchEx.exec(line)) !== null) {
      if (match.index === matchEx.lastIndex) {
        matchEx.lastIndex++;
      }

      numbersInString.push(task === 1 ? match[0] : match[1]);
    }

    const numberString =
      String(wordToNumber(numbersInString.slice(0)[0])) +
      String(wordToNumber(numbersInString.slice(-1)[0]));

    result += parseInt(numberString);
  }

  return result.toString();
}

export default day1;

File: 2023/kennydl/cmd/solver/day1/day1.go
package day1

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func Run() {
	file, _ := os.Open("cmd/solver/day1/input.txt")
	defer file.Close()

	partOne(file)
	partTwo(file)
}

func partOne(file *os.File) {
	scanner := bufio.NewScanner(file)
	re, _ := regexp.Compile(`(\d)`)
	result := 0
	
	for scanner.Scan()	{
		line := scanner.Text()
		match := re.FindAllString(line, -1)
		if match != nil {
			value, _ := strconv.Atoi(getNumber(line, re))
			result += value
		}
	}

	fmt.Println("part 1:", result)
}

var numberValues = map[string]string{
	"one": "1",
	"two": "2",
	"three": "3",
	"four": "4",
	"five": "5",
	"six" : "6",
	"seven": "7",
	"eight": "8",
	"nine": "9"}

func partTwo(file *os.File) {
	file.Seek(0, 0)
	scanner := bufio.NewScanner(file)
	pattern := `(\d)`
	for k := range numberValues {
		pattern = pattern + `|(` + k + `)`
	}
	re, _ := regexp.Compile(pattern)
	result := 0

	for scanner.Scan()	{
		line := scanner.Text()
		value, _ := strconv.Atoi(getNumber(line, re))
		result += value
	}

	fmt.Println("part 2:", result)
}

func getNumber(line string, re *regexp.Regexp) string  {
	firstDigit, lastDigit, len := "", "", len(line)
	for i := range line {
		firstValue, lastValue := re.FindString(line[0:i+1]), re.FindString(line[len-1-i:len])
		if firstValue != "" && firstDigit == "" {
			firstDigit = convertToDigitString(firstValue)
		}
		if lastValue != "" && lastDigit == "" {
			lastDigit = convertToDigitString(lastValue)
		}
		if firstDigit != "" && lastDigit != "" {
			return firstDigit + lastDigit
		}
	}
	return ""
}

func convertToDigitString(value string) string  {
	digit := value
	_, err := strconv.Atoi(digit)

	if(err == nil) {
		return digit
	}
	return numberValues[value]
}